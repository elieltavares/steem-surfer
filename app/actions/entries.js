import {
  getDiscussions,
  getRepliesByLastUpdate
} from '../backend/steem-client';

export const FETCH_BEGIN = 'entries/FETCH_BEGIN';
export const FETCH_OK = 'entries/FETCH_OK';
export const FETCH_ERROR = 'entries/FETCH_ERROR';
export const INVALIDATE = 'entries/INVALIDATE';
export const SET_READ = 'entries/SET_READ';
export const SET_VOTED = 'entries/SET_VOTED';
export const UPDATE_ENTRY = 'entries/UPDATE';

export const makeGroupKeyForEntries = (what, tag = '') => {
  if (tag) {
    return `${what}-${tag}`;
  }
  return `${what}`;
};

export function fetchEntries(what, tag = '', more = false) {
  return (dispatch, getState) => {
    const { entries } = getState();
    const pageSize = 20;

    const groupKey = makeGroupKeyForEntries(what, tag);

    if (!more && entries.getIn([groupKey, 'entries']).size) {
      return;
    }

    const lastEntry = entries
      .getIn([groupKey, 'entries'])
      .valueSeq()
      .last();

    let fn;
    let query;

    if (tag.startsWith('@')) {
      const username = tag.replace('@', '');
      switch (what) {
        case 'feed':
          fn = getDiscussions;
          query = {
            tag: username,
            limit: pageSize,
            start_author: undefined,
            start_permlink: undefined
          };
          break;
        case 'comments':
          fn = getDiscussions;
          query = {
            limit: pageSize,
            start_author: username,
            start_permlink: undefined
          };
          break;
        case 'replies':
          fn = getRepliesByLastUpdate;
          query = {
            start_author: username,
            start_permlink: undefined,
            limit: pageSize
          };
          break;
        default:
          fn = getDiscussions;
          query = {
            tag: username,
            limit: pageSize,
            start_author: undefined,
            start_permlink: undefined
          };
          break;
      }
    } else {
      fn = getDiscussions;
      // make sure tag is not null or undefined. it should be empty string.
      query = {
        tag: tag || '',
        limit: pageSize,
        start_author: undefined,
        start_permlink: undefined
      };
    }

    if (lastEntry) {
      query.start_author = lastEntry.author;
      query.start_permlink = lastEntry.permlink;
    }

    dispatch({
      type: FETCH_BEGIN,
      payload: { group: groupKey }
    });

    const fnArgs = what === 'replies' ? [query] : [what, query];

    fn(...fnArgs)
      .then(resp => {
        dispatch({
          type: FETCH_OK,
          payload: {
            data: resp,
            group: groupKey,
            hasMore: resp.length >= pageSize
          }
        });

        return resp;
      })
      .catch(e => {
        dispatch({
          type: FETCH_ERROR,
          payload: { group: groupKey, error: e }
        });
      });
  };
}

export function updateEntry(newData) {
  return dispatch => {
    dispatch({
      type: UPDATE_ENTRY,
      payload: { data: newData }
    });
  };
}

export function invalidateEntries(what, tag = '') {
  return dispatch => {
    const groupKey = makeGroupKeyForEntries(what, tag);

    dispatch({
      type: INVALIDATE,
      payload: { group: groupKey }
    });
  };
}
