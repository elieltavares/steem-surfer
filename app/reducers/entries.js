import { Record, Map, OrderedMap } from 'immutable';
import { LOCATION_CHANGE } from 'react-router-redux';
import {
  FETCH_BEGIN,
  FETCH_OK,
  FETCH_ERROR,
  INVALIDATE,
  UPDATE_ENTRY,
  makeGroupKeyForEntries
} from '../actions/entries';

import filters from '../constants/filters.json';

export const EntryGroupRecord = Record({
  entries: OrderedMap({}),
  err: null,
  loading: false,
  hasMore: true
});

const defaultState = Map();

export default function entries(state = defaultState, action) {
  switch (action.type) {
    case LOCATION_CHANGE: {
      // Create entry group when location changed
      const path = action.payload.pathname.split('/');
      if (path.length > 0 && filters.includes(path[1])) {
        const filter = path[1];
        const tag = path[2] || null;

        const groupKey = makeGroupKeyForEntries(filter, tag);
        if (state.get(groupKey) === undefined) {
          return state.set(groupKey, new EntryGroupRecord());
        }
      }

      if (path.length >= 1 && path[1].startsWith('@')) {
        const filter = path[2] || 'blog';
        const tag = path[1];

        if (!['blog', 'comments', 'replies', 'feed'].includes(filter)) {
          return state;
        }

        const groupKey = makeGroupKeyForEntries(filter, tag);
        if (state.get(groupKey) === undefined) {
          return state.set(groupKey, new EntryGroupRecord());
        }
      }

      return state;
    }
    case FETCH_BEGIN: {
      const groupKey = action.payload.group;

      return state
        .setIn([groupKey, 'err'], null)
        .setIn([groupKey, 'loading'], true);
    }
    case FETCH_OK: {
      const { group: groupKey, data: newEntries, hasMore } = action.payload;

      let newState = state
        .setIn([groupKey, 'err'], null)
        .setIn([groupKey, 'loading'], false)
        .setIn([groupKey, 'hasMore'], hasMore);

      newEntries.forEach(entry => {
        if (!newState.hasIn([groupKey, 'entries', `${entry.author}-${entry.permlink}`])) {
          newState = newState.setIn(
            [groupKey, 'entries', `${entry.author}-${entry.permlink}`],
            entry
          );
        }
      });

      return newState;
    }
    case FETCH_ERROR: {
      const groupKey = action.payload.group;
      return state
        .setIn([groupKey, 'err'], action.payload.error)
        .setIn([groupKey, 'loading'], false);
    }
    case INVALIDATE: {
      const groupKey = action.payload.group;
      return state
        .setIn([groupKey, 'entries'], OrderedMap({}))
        .setIn([groupKey, 'err'], null)
        .setIn([groupKey, 'loading'], false);
    }
    case UPDATE_ENTRY: {
      const { data } = action.payload;
      let newState = state.asImmutable();

      newState.keySeq().forEach(groupKey => {
        if (newState.hasIn([groupKey, 'entries', `${data.author}-${data.permlink}`])) {
          newState = newState.setIn([groupKey, 'entries', `${data.author}-${data.permlink}`], data);
        }
      });

      return newState;
    }
    default:
      return state;
  }
}
