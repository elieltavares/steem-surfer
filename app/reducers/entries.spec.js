/* eslint-disable */
import { Record, Map, OrderedMap } from 'immutable';

import entries from './entries';
import { EntryGroupRecord } from './entries';
import { LOCATION_CHANGE } from 'react-router-redux';

import {
  FETCH_BEGIN,
  FETCH_OK,
  FETCH_ERROR,
  INVALIDATE
} from '../actions/entries';

import deepFreeze from 'deep-freeze';

describe('entries reducer', () => {
  it('(1) should handle initial state', () => {
    expect(entries(undefined, {})).toMatchSnapshot();
  });

  it('(2) should create group key according to location. action:LOCATION_CHANGE', () => {
    const stateBefore = Map();

    const action = {
      type: LOCATION_CHANGE,
      payload: {
        pathname: '/trending'
      }
    };

    const res = entries(stateBefore, action);
    expect(res).toMatchSnapshot();
  });

  it('(3) should create group key according to location. action:LOCATION_CHANGE', () => {
    const stateBefore = Map({
      trending: new EntryGroupRecord({
        entries: OrderedMap({}),
        err: null,
        loading: false
      })
    });

    const action = {
      type: LOCATION_CHANGE,
      payload: {
        pathname: '/trending/art'
      }
    };
    const res = entries(stateBefore, action);
    expect(res).toMatchSnapshot();
  });

  it('(4) should  not create group for non-filter path. action:LOCATION_CHANGE', () => {
    const stateBefore = Map({
      trending: new EntryGroupRecord({
        entries: OrderedMap({}),
        err: null,
        loading: false
      }),
      'trending-art': new EntryGroupRecord({
        entries: OrderedMap({}),
        err: null,
        loading: false
      })
    });

    const action = {
      type: LOCATION_CHANGE,
      payload: {
        pathname: '/lipsum/art'
      }
    };
    const res = entries(stateBefore, action);
    expect(res).toMatchSnapshot();
  });

  it('(5) start fetching "trending". action:FETCH_BEGIN', () => {
    const stateBefore = Map({
      trending: new EntryGroupRecord({
        entries: OrderedMap({}),
        err: null,
        loading: false
      }),
      'trending-art': new EntryGroupRecord({
        entries: OrderedMap({}),
        err: null,
        loading: false
      })
    });

    deepFreeze(stateBefore);

    const action = {
      type: FETCH_BEGIN,
      payload: {
        group: 'trending'
      }
    };

    const res = entries(stateBefore, action);
    expect(res).toMatchSnapshot();
  });

  it('(6) start fetching "trending". should keep old entries. action:FETCH_BEGIN', () => {
    const stateBefore = Map({
      trending: new EntryGroupRecord({
        entries: OrderedMap({ 34: { body: 'foo' }, 12: { body: 'bar' } }),
        err: null,
        loading: false
      }),
      'trending-art': new EntryGroupRecord({
        entries: OrderedMap({}),
        err: null,
        loading: false
      })
    });

    deepFreeze(stateBefore);

    const action = {
      type: FETCH_BEGIN,
      payload: {
        group: 'trending'
      }
    };

    const res = entries(stateBefore, action);
    expect(res).toMatchSnapshot();
  });

  it('(7) fetching "trending" completed. action:FETCH_OK', () => {
    const stateBefore = Map({
      trending: new EntryGroupRecord({
        entries: OrderedMap({}),
        err: null,
        loading: true
      }),
      'trending-art': new EntryGroupRecord({
        entries: OrderedMap({}),
        err: null,
        loading: false
      })
    });

    deepFreeze(stateBefore);

    const action = {
      type: FETCH_OK,
      payload: {
        data: [
          {
            id: 14,
            author: 'good-karma',
            permlink: 'welcome to esteem',
            votes: 5
          },
          { id: 12, author: 'talhasch', permlink: 'lorem ipsum', votes: 12 },
          { id: 16, author: 'chrisbolten', permlink: 'sit amet', votes: 3 }
        ],
        group: 'trending'
      }
    };

    const res = entries(stateBefore, action);
    expect(res).toMatchSnapshot();
  });

  it('(8) should not add exiting item to entries. action:FETCH_OK', () => {
    const stateBefore = Map({
      trending: new EntryGroupRecord({
        entries: OrderedMap({
          '12': {
            id: 12,
            author: 'good-karma',
            permlink: 'welcome to esteem',
            votes: 5
          }
        }),
        err: null,
        loading: true
      })
    });

    deepFreeze(stateBefore);

    const action = {
      type: FETCH_OK,
      payload: {
        data: [
          {
            id: 12,
            author: 'good-karma',
            permlink: 'welcome to esteem',
            votes: 5
          }
        ],
        group: 'trending'
      }
    };

    const res = entries(stateBefore, action);
    expect(res).toMatchSnapshot();
  });

  it('(9) simulate load more. action:FETCH_OK', () => {
    let stateBefore = Map({
      trending: new EntryGroupRecord({
        entries: OrderedMap({}),
        err: null,
        loading: true
      }),
      'trending-art': new EntryGroupRecord({
        entries: OrderedMap({}),
        err: null,
        loading: false
      })
    });

    stateBefore = stateBefore.setIn(['trending', 'entries', '14'], {
      id: 14,
      author: 'good-karma',
      permlink: 'welcome to esteem',
      votes: 5
    });
    stateBefore = stateBefore.setIn(['trending', 'entries', '12'], {
      id: 12,
      author: 'talhasch',
      permlink: 'lorem ipsum',
      votes: 12
    });

    deepFreeze(stateBefore);

    const action = {
      type: FETCH_OK,
      payload: {
        data: [
          { id: 16, author: 'chrisbolten', permlink: 'sit amet', votes: 3 },
          { id: 3, author: 'dunsky', permlink: 'lorem', votes: 31 }
        ],
        group: 'trending'
      }
    };

    const res = entries(stateBefore, action);
    expect(res).toMatchSnapshot();
  });

  it('(10) error occurred while fetching "trending". action:FETCH_ERROR', () => {
    let stateBefore = Map({
      trending: new EntryGroupRecord({
        entries: OrderedMap({}),
        err: null,
        loading: true
      }),
      'trending-art': new EntryGroupRecord({
        entries: OrderedMap({}),
        err: null,
        loading: false
      })
    });

    stateBefore = stateBefore.setIn(['trending', 'entries', '14'], {
      id: 14,
      author: 'good-karma',
      permlink: 'welcome to esteem',
      votes: 5
    });

    deepFreeze(stateBefore);

    const action = {
      type: FETCH_ERROR,
      payload: {
        group: 'trending',
        error: 'an error has occurred'
      }
    };

    const res = entries(stateBefore, action);
    expect(res).toMatchSnapshot();
  });

  it('(11) invalidate "trending". action:INVALIDATE', () => {
    let stateBefore = Map({
      trending: new EntryGroupRecord({
        entries: OrderedMap({}),
        err: null,
        loading: false
      }),
      'trending-art': new EntryGroupRecord({
        entries: OrderedMap({}),
        err: null,
        loading: false
      })
    });

    stateBefore = stateBefore.setIn(['trending', 'entries', '14'], {
      id: 14,
      author: 'good-karma',
      permlink: 'welcome to esteem',
      votes: 5
    });

    deepFreeze(stateBefore);

    const action = {
      type: INVALIDATE,
      payload: {
        group: 'trending'
      }
    };

    const res = entries(stateBefore, action);
    expect(res).toMatchSnapshot();
  });

  it('(12) should reset error. action:FETCH_BEGIN', () => {
    let stateBefore = Map({
      trending: new EntryGroupRecord({
        entries: OrderedMap({}),
        err: 'an error',
        loading: false
      })
    });

    deepFreeze(stateBefore);

    const action = {
      type: FETCH_BEGIN,
      payload: {
        group: 'trending'
      }
    };

    const res = entries(stateBefore, action);
    expect(res).toMatchSnapshot();
  });
});
