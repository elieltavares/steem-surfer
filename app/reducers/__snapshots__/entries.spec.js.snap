// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`entries reducer (1) should handle initial state 1`] = `Immutable.Map {}`;

exports[`entries reducer (2) should create group key according to location. action:LOCATION_CHANGE 1`] = `
Immutable.Map {
  "trending": Immutable.Record {
    "entries": Immutable.OrderedMap {},
    "err": null,
    "loading": false,
    "hasMore": true,
  },
}
`;

exports[`entries reducer (3) should create group key according to location. action:LOCATION_CHANGE 1`] = `
Immutable.Map {
  "trending": Immutable.Record {
    "entries": Immutable.OrderedMap {},
    "err": null,
    "loading": false,
    "hasMore": true,
  },
  "trending-art": Immutable.Record {
    "entries": Immutable.OrderedMap {},
    "err": null,
    "loading": false,
    "hasMore": true,
  },
}
`;

exports[`entries reducer (4) should  not create group for non-filter path. action:LOCATION_CHANGE 1`] = `
Immutable.Map {
  "trending": Immutable.Record {
    "entries": Immutable.OrderedMap {},
    "err": null,
    "loading": false,
    "hasMore": true,
  },
  "trending-art": Immutable.Record {
    "entries": Immutable.OrderedMap {},
    "err": null,
    "loading": false,
    "hasMore": true,
  },
}
`;

exports[`entries reducer (5) start fetching "trending". action:FETCH_BEGIN 1`] = `
Immutable.Map {
  "trending": Immutable.Record {
    "entries": Immutable.OrderedMap {},
    "err": null,
    "loading": true,
    "hasMore": true,
  },
  "trending-art": Immutable.Record {
    "entries": Immutable.OrderedMap {},
    "err": null,
    "loading": false,
    "hasMore": true,
  },
}
`;

exports[`entries reducer (6) start fetching "trending". should keep old entries. action:FETCH_BEGIN 1`] = `
Immutable.Map {
  "trending": Immutable.Record {
    "entries": Immutable.OrderedMap {
      "12": Object {
        "body": "bar",
      },
      "34": Object {
        "body": "foo",
      },
    },
    "err": null,
    "loading": true,
    "hasMore": true,
  },
  "trending-art": Immutable.Record {
    "entries": Immutable.OrderedMap {},
    "err": null,
    "loading": false,
    "hasMore": true,
  },
}
`;

exports[`entries reducer (7) fetching "trending" completed. action:FETCH_OK 1`] = `
Immutable.Map {
  "trending": Immutable.Record {
    "entries": Immutable.OrderedMap {
      "good-karma-welcome to esteem": Object {
        "author": "good-karma",
        "id": 14,
        "permlink": "welcome to esteem",
        "votes": 5,
      },
      "talhasch-lorem ipsum": Object {
        "author": "talhasch",
        "id": 12,
        "permlink": "lorem ipsum",
        "votes": 12,
      },
      "chrisbolten-sit amet": Object {
        "author": "chrisbolten",
        "id": 16,
        "permlink": "sit amet",
        "votes": 3,
      },
    },
    "err": null,
    "loading": false,
    "hasMore": undefined,
  },
  "trending-art": Immutable.Record {
    "entries": Immutable.OrderedMap {},
    "err": null,
    "loading": false,
    "hasMore": true,
  },
}
`;

exports[`entries reducer (8) should not add exiting item to entries. action:FETCH_OK 1`] = `
Immutable.Map {
  "trending": Immutable.Record {
    "entries": Immutable.OrderedMap {
      "12": Object {
        "author": "good-karma",
        "id": 12,
        "permlink": "welcome to esteem",
        "votes": 5,
      },
      "good-karma-welcome to esteem": Object {
        "author": "good-karma",
        "id": 12,
        "permlink": "welcome to esteem",
        "votes": 5,
      },
    },
    "err": null,
    "loading": false,
    "hasMore": undefined,
  },
}
`;

exports[`entries reducer (9) simulate load more. action:FETCH_OK 1`] = `
Immutable.Map {
  "trending": Immutable.Record {
    "entries": Immutable.OrderedMap {
      "14": Object {
        "author": "good-karma",
        "id": 14,
        "permlink": "welcome to esteem",
        "votes": 5,
      },
      "12": Object {
        "author": "talhasch",
        "id": 12,
        "permlink": "lorem ipsum",
        "votes": 12,
      },
      "chrisbolten-sit amet": Object {
        "author": "chrisbolten",
        "id": 16,
        "permlink": "sit amet",
        "votes": 3,
      },
      "dunsky-lorem": Object {
        "author": "dunsky",
        "id": 3,
        "permlink": "lorem",
        "votes": 31,
      },
    },
    "err": null,
    "loading": false,
    "hasMore": undefined,
  },
  "trending-art": Immutable.Record {
    "entries": Immutable.OrderedMap {},
    "err": null,
    "loading": false,
    "hasMore": true,
  },
}
`;

exports[`entries reducer (10) error occurred while fetching "trending". action:FETCH_ERROR 1`] = `
Immutable.Map {
  "trending": Immutable.Record {
    "entries": Immutable.OrderedMap {
      "14": Object {
        "author": "good-karma",
        "id": 14,
        "permlink": "welcome to esteem",
        "votes": 5,
      },
    },
    "err": "an error has occurred",
    "loading": false,
    "hasMore": true,
  },
  "trending-art": Immutable.Record {
    "entries": Immutable.OrderedMap {},
    "err": null,
    "loading": false,
    "hasMore": true,
  },
}
`;

exports[`entries reducer (11) invalidate "trending". action:INVALIDATE 1`] = `
Immutable.Map {
  "trending": Immutable.Record {
    "entries": Immutable.OrderedMap {},
    "err": null,
    "loading": false,
    "hasMore": true,
  },
  "trending-art": Immutable.Record {
    "entries": Immutable.OrderedMap {},
    "err": null,
    "loading": false,
    "hasMore": true,
  },
}
`;

exports[`entries reducer (12) should reset error. action:FETCH_BEGIN 1`] = `
Immutable.Map {
  "trending": Immutable.Record {
    "entries": Immutable.OrderedMap {},
    "err": null,
    "loading": true,
    "hasMore": true,
  },
}
`;
